src/utils/daos.c

```diff
@@ -0,0 +1,870 @@
+			break;
+		case 'd':
+			ap->path_str = acopy_str(optarg);
+			break;
+		case 't':
+			ap->type_str = acopy_str(optarg);
+			break;
+		case 'o':
+			ap->oclass_str = acopy_str(optarg);
+			break;
+		case 'z':
+			ap->chunk_size = tobytes(optarg);
+			if (ap->chunk_size == 0 ||
+			    (ap->chunk_size == ULLONG_MAX && errno != 0)) {
+				fprintf(stderr, "failed to parse chunk_size:"
+					"%s\n", optarg);
+				return RC_NO_HELP;
+			}
+			break;
+		case 's':
+			ap->snapname_str = acopy_str(optarg);
+			break;
+		case 'e':
+			ap->epc = strtoull(optarg, NULL, 10);
+			if (ap->epc == 0 ||
+			    (ap->epc == ULLONG_MAX && errno != 0)) {
+				fprintf(stderr, "failed to parse epc: %s\n",
+					optarg);
+				return RC_NO_HELP;
+			}
+			break;
+		case 'r':
+			ap->epcrange_str = acopy_str(optarg);
+			rc = epoch_range_parse(ap);
+			if (rc != 0) {
+				fprintf(stderr, "failed to parse epcrange\n");
+				return RC_NO_HELP;
+			}
+			break;
+		default:
+			printf("unknown option : %d\n", rc);
+			return RC_PRINT_HELP;
+		}
+	}
+
+	cmd_args_print(ap);
+
+	/* Check for any unimplemented commands, print help */
+	if (ap->p_op != -1 &&
+	    (ap->p_op == POOL_LIST_CONTAINERS ||
+	     ap->p_op == POOL_STAT ||
+	     ap->p_op == POOL_GET_PROP ||
+	     ap->p_op == POOL_GET_ATTR ||
+	     ap->p_op == POOL_LIST_ATTRS)) {
+		fprintf(stderr,
+			"pool %s not yet implemented\n", cmdname);
+		return RC_NO_HELP;
+	}
+
+	if (ap->c_op != -1 &&
+	    (ap->c_op == CONT_LIST_OBJS ||
+	     ap->c_op == CONT_QUERY ||
+	     ap->c_op == CONT_STAT ||
+	     ap->c_op == CONT_GET_PROP ||
+	     ap->c_op == CONT_SET_PROP ||
+	     ap->c_op == CONT_LIST_ATTRS ||
+	     ap->c_op == CONT_DEL_ATTR ||
+	     ap->c_op == CONT_GET_ATTR ||
+	     ap->c_op == CONT_SET_ATTR ||
+	     ap->c_op == CONT_CREATE_SNAP ||
+	     ap->c_op == CONT_LIST_SNAPS ||
+	     ap->c_op == CONT_DESTROY_SNAP ||
+	     ap->c_op == CONT_ROLLBACK)) {
+		fprintf(stderr,
+			"container %s not yet implemented\n", cmdname);
+		return RC_NO_HELP;
+	}
+
+	/* Check the pool UUID (required) */
+	if (uuid_is_null(ap->pool_uuid)) {
+		fprintf(stderr, "pool UUID required\n");
+		fflush(stderr);
+		return RC_PRINT_HELP;
+	}
+
+	/* Check the pool service ranks.(required) */
+	if (ap->mdsrv_str == NULL) {
+		fprintf(stderr, "--svc must be specified\n");
+		fflush(stderr);
+		return RC_PRINT_HELP;
+	}
+
+	if (ap->mdsrv == NULL) {
+		fprintf(stderr, "failed to parse --svc=%s\n", ap->mdsrv_str);
+		return RC_NO_HELP;
+	}
+
+	if (ap->mdsrv->rl_nr == 0) {
+		fprintf(stderr, "--svc must not be empty\n");
+		fflush(stderr);
+		rc = RC_PRINT_HELP;
+		goto bad_opt_free_mdsrv;
+	}
+
+	/* TODO: decide if for container operations the code should
+	 * check that a container UUID or path has been provided,
+	 * and if a path, that container UUID can be looked up (UNS).
+	 * Or, if that checking should be in container command handling.
+	 */
+	return 0;
+
+bad_opt_free_mdsrv:
+	daos_rank_list_free(ap->mdsrv);
```

Comment:
you have to free this even in success condition..

Reply:
agree, will fix in next patch (won't have bad_ in the label name)